### PR TITLE
call_linker now returns exit_code for better user response on linking_error, fixes #7141

### LINKED ISSUE #7141: non informative link time error when the c compiler couldn't run
**Original bug ID:** 7141
**Reporter:** @sliquister
**Status:** acknowledged (set by @damiendoligez on 2016-02-25T13:24:13Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Version:** 4.02.3
**Category:** compiler driver
**Tags:** junior_job
**Monitored by:** @diml @hcarty

## Bug description

When failing to run the c compiler at link time (in my case because
the command line is too big), ocamlopt just fails with this error:

$ /tmp/script.sh 
File "caml_startup", line 1:
Error: Error during linking

which is not super helpful. Even after adding -verbose, we still
didn't know what was failing and strac'ing was necessary.

This is because ocamlopt runs [Sys.command .. = 0], presumably relying
on the command to report an error. I think it would be preferable if
the Linking_error exception thrown in the false case could contain the
exit code, and perhaps the program name, so the error could look like:

$ /tmp/script.sh 
File "caml_startup", line 1:
Error: Error during linking (exit code 127 from running g++)

## Steps to reproduce

cd /tmp
rm -f a.cmx* b.cmx*
touch a.ml b.ml
ocamlopt -c a.ml
ocamlopt -ccopt "$(printf "%70000s")" -a a.cmx -o a.cmxa
ocamlopt -c b.ml
ocamlopt -ccopt "$(printf "%70000s")" -a b.cmx -o b.cmxa
ocamlopt a.cmxa b.cmxa -o out

### PR DESCRIPTION
**Pull request by Anukriti12**
Created additional error `Linking_error_c_compiler` fixes #7141

### CODE REVIEW

File: asmcomp/asmlink.ml [hunk 1]
```diff
@@ -334,7 +335,7 @@ let call_linker file_list startup_file output_name =
     else Ccomp.Exe
   in
   if not (Ccomp.call_linker mode output_name files c_lib)
-  then raise(Error Linking_error)
+  then raise(Error Linking_error_c_compiler)
```
Comment (gasche) [inline asmcomp/asmlink.ml]: The error message says that the exit code is `127` and the caller program is `g++`, but the way you raise the exception does not ensure this: `call_linker` may return `false` even with other exit codes, and we don't know which linker was called.

**Review by gasche — COMMENTED** [review]
I think the approach in this PR is not completely right: it assumes that the exit code is always `127` and the linker called always `g++`, but this is not always the case. I would propose to:

- Focus on the exit code only: have a working PR that prints the exit code in the error message, and forget about the linker program name for now; we can always add this in a later PR.

- Change `Ccomp.call_linker` to return the exit code of the call, instead of just a boolean (this means that all callers have to be adapted to check `= 0` themselves)

- Add the exit code number as a parameter of the exception: `exception Linking_error of int`. (Note: instead of defining a new exception, we just extend `Linking_error`.)

**Comment by v-gb** [comment]
In principle, https://github.com/ocaml/ocaml/pull/1727 should have fixed the problem. I still see a bad error my machine, but it looks like a glibc bug where system() returns wrong values (the child gets E2BIG in exec, which is errno 7, which gets reinterpreted as status 7, which is signalled by sibgus, which Sys.command returns as 255).

**Comment by gasche** [comment]
@sliquister ah, I had missed this change. (cc @trefis in case). Do you think the change in the present PR, which adds the exit code to the non-127 linking-error case, would possibly be helpful for users?

**Comment by v-gb** [comment]
The only confusing case I have ever seen is with exit code 127 (because that's what system uses whenever anything goes wrong), because the rest of the time the program should print an error. It might be useful if the linker was killed by a signal, but I haven't tried seeing if any error is printed then.

**Comment by Anukriti12** [comment]
@gasche Can you please review once?

File: bytecomp/bytelink.ml [hunk 2]
```diff
@@ -577,7 +577,7 @@ let build_custom_runtime prim_name exec_name =
       [] in
   Ccomp.call_linker Ccomp.Exe exec_name
     (debug_prefix_map @ [prim_name] @ List.rev !Clflags.ccobjs @ [runtime_lib])
-    (Clflags.std_include_flag "-I" ^ " " ^ Config.bytecomp_c_libraries)
+    (Clflags.std_include_flag "-I" ^ " " ^ Config.bytecomp_c_libraries) = 0
```
Comment (gasche) [inline bytecomp/bytelink.ml]: I would prefer a `let exitcode` binding here for readability, and same thing in the other diff below.

File: asmcomp/asmlink.ml [hunk 3]
```diff
@@ -333,8 +334,9 @@ let call_linker file_list startup_file output_name =
     else if !Clflags.output_c_object then Ccomp.Partial
     else Ccomp.Exe
   in
-  if not (Ccomp.call_linker mode output_name files c_lib)
-  then raise(Error Linking_error)
+  let exitcode = Ccomp.call_linker mode output_name files c_lib in
+  if not (exitcode = 0)
+  then raise(Error(Linking_error (exitcode)))
```
Comment (gasche) [inline asmcomp/asmlink.ml]: The parentheses around `exitcode` are unnecessary and using spacing for functoin calls is more idiomatic. I would reformulate this (and above) as

```
then raise (Error (Linking_error exitcode))
```

**Review by gasche — APPROVED** [review]
From my reading of the reports on this error, the exit code is not often necessary but it may still be useful in some cases (including the tool being buggy and returning something else than 127 when it should): it does not hurt and can help. I am in favor in merging if the style-small style-level nitpicks I made are adressed.

**Pull request merged**